Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp

```diff
@@ -351,58 +351,70 @@ auto MediaStreamTrack::takePhoto(PhotoSettings&& settings) -> Ref<TakePhotoPromi
 
-            ALWAYS_LOG(identifier, "resolving promise");
-            promise.resolve(WTFMove(*result.capabilities));
-        });
+    PhotoCapabilitiesPromise::Producer producer;
+    Ref<PhotoCapabilitiesPromise> promise = producer;
+
+    m_private->getPhotoCapabilities()->whenSettled(RunLoop::main(),
+        [protectedThis = Ref { *this }, producer = WTFMove(producer)] (auto&& result) mutable {
+
+        // https://w3c.github.io/mediacapture-image/#ref-for-dom-imagecapture-getphotocapabilities②
+        // If the data cannot be gathered for any reason (for example, the MediaStreamTrack being ended
+        // asynchronously), then reject p with a new DOMException whose name is OperationError, and
+        // abort these steps.
+        if (!result) {
+            producer.reject(Exception { ExceptionCode::UnknownError, WTFMove(result.error()) });
+            return;
+        }
+
+        if (RefPtr context = protectedThis->scriptExecutionContext(); !context || context->activeDOMObjectsAreStopped() || protectedThis->m_ended) {
+            producer.reject(Exception { ExceptionCode::OperationError, "Track has ended"_s });
+            return;
+        }
+
+        producer.resolve(WTFMove(result.value()));
     });
+
+    return promise;
 }
 
-void MediaStreamTrack::getPhotoSettings(DOMPromiseDeferred<IDLDictionary<PhotoSettings>>&& promise)
+auto MediaStreamTrack::getPhotoSettings() -> Ref<PhotoSettingsPromise>
 {
-    auto identifier = LOGIDENTIFIER;
+    if (m_ended)
+        return PhotoSettingsPromise::createAndReject(Exception { ExceptionCode::InvalidStateError, "Track has ended"_s });
 
-    m_private->getPhotoSettings()->whenSettled(RunLoop::main(), [this, protectedThis = Ref { *this }, promise = WTFMove(promise), identifier = WTFMove(identifier)] (auto&& result) mutable {
-        queueTaskKeepingObjectAlive(*this, TaskSource::ImageCapture, [this, promise = WTFMove(promise), result = WTFMove(result), identifier = WTFMove(identifier)] () mutable {
-            // https://w3c.github.io/mediacapture-image/#ref-for-dom-imagecapture-getphotosettings②
-            // If the data cannot be gathered for any reason (for example, the MediaStreamTrack being ended
-            // asynchronously), then reject p with a new DOMException whose name is OperationError, and
-            // abort these steps.
-            if (!result) {
-                ERROR_LOG(identifier, "rejecting promise: ", result.error());
-                promise.reject(Exception { ExceptionCode::OperationError, WTFMove(result.error()) });
-                return;
-            }
-            if (m_readyState != State::Live) {
-                ERROR_LOG(identifier, "rejecting promise, track has ended");
-                promise.reject(Exception { ExceptionCode::OperationError, "Track has ended"_s });
-                return;
-            }
+    PhotoSettingsPromise::Producer producer;
+    Ref<PhotoSettingsPromise> promise = producer;
```

Comment:
same here, the producer isn't necessary.
return m_private->getPhotoSettings()->whenSettled(queue, [...] (auto&& result) { return PhotoSettinedPromise::createAndResolve...)} );